cilium/socket_option_ip_transparent.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <asm-generic/socket.h>
+#include <netinet/in.h>
+
+#include <cerrno>
+#include <cstdint>
+#include <vector>
+
+#include "envoy/config/core/v3/socket_option.pb.h"
+#include "envoy/network/address.h"
+#include "envoy/network/socket.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/common/utility.h"
+
+#include "absl/types/optional.h"
+#include "cilium/privileged_service_client.h"
+
+namespace Envoy {
+namespace Cilium {
+
+// Socket Option that sets the socket option IP_TRANSPARENT (IPV6_TRANSPARENT) on the socket.
+// It uses the Cilium Privileged Service to call out to the starter process to do the actual
+// privileged syscall - as the Envoy process itself doesn't have the required capabilities.
+class IpTransparentSocketOption : public Network::Socket::Option,
+                                  public Logger::Loggable<Logger::Id::filter> {
+public:
+  IpTransparentSocketOption() { ENVOY_LOG(debug, "Cilium IpTransparentSocketOption()"); }
+
+  absl::optional<Network::Socket::Option::Details>
+  getOptionDetails(const Network::Socket&,
+                   envoy::config::core::v3::SocketOption::SocketState) const override {
+    return absl::nullopt;
+  }
+
+  bool setOption(Network::Socket& socket,
+                 envoy::config::core::v3::SocketOption::SocketState state) const override {
+    // Only set the option once per socket
+    if (state != envoy::config::core::v3::SocketOption::STATE_PREBIND) {
+      ENVOY_LOG(trace, "Skipping setting socket ({}) option IP_TRANSPARENT, state != STATE_PREBIND",
+                socket.ioHandle().fdDoNotUse());
+      return true;
+    }
+
+    auto& cilium_calls = PrivilegedService::Singleton::get();
+
+    auto ipVersion = socket.ipVersion();
+    if (!ipVersion.has_value()) {
+      ENVOY_LOG(critical, "Socket address family is not available, can not choose source address");
+      return false;
+    }
+
+    uint32_t one = 1;
+
+    // Set ip transparent option based on the socket address family
+    auto ip_transparent_socket_option = IP_TRANSPARENT;
+    auto ip_transparent_socket_option_name = "IP_TRANSPARENT";
+    if (*ipVersion == Network::Address::IpVersion::v6) {
+      ip_transparent_socket_option = IPV6_TRANSPARENT;
+      ip_transparent_socket_option_name = "IPV6_TRANSPARENT";
+    }
+
+    auto status = cilium_calls.setsockopt(socket.ioHandle().fdDoNotUse(), SOL_IP,
```

Comment:
Should this be `SOL_IPV6` for IPv6?

Reply:
of course - even though this was still "fixed" on `main` it was still wrong on my working copy where i copied this from. thanks a lot!